middleware/ratelimiter/store.go

```diff
@@ -0,0 +1,109 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/System-Glitch/goyave/v3"
+)
+
+type limiter struct {
+	requestQuota  int
+	counter       int
+	quotaDuration time.Duration
+	resetsAt      time.Time
+	mx            sync.RWMutex
+}
+
+func newLimiter(requestQuota int, quotaDuration time.Duration) *limiter {
+	return &limiter{
+		requestQuota:  requestQuota,
+		quotaDuration: quotaDuration,
+		counter:       0,
+		resetsAt:      time.Now().Add(quotaDuration),
+	}
+}
+
+func (l *limiter) validateAndUpdate(response *goyave.Response) bool {
+
+	l.mx.RLock()
+	defer l.mx.RUnlock()
+
+	if l.hasExpired() {
+		l.reset()
+	} else if l.hasExceededRequestQuota() {
+		l.updateResponseHeaders(response)
+		return false
+	}
+
+	l.increment()
+	l.updateResponseHeaders(response)
+	return true
+}
+
+func (l *limiter) updateResponseHeaders(response *goyave.Response) {
+	response.Header().Set(
+		"RateLimit-Limit",
+		fmt.Sprintf("%v, %v;w=%v", l.requestQuota, l.requestQuota, l.quotaDuration.Seconds()),
+	)
+
+	response.Header().Set(
+		"RateLimit-Remaining",
+		fmt.Sprintf("%v", l.getRemainingRequestQuota()),
+	)
+
+	response.Header().Set(
+		"RateLimit-Reset",
+		fmt.Sprintf("%v", l.getSecondsToQuotaReset()),
+	)
+}
+
+func (l *limiter) reset() {
+	l.resetsAt = time.Now().Add(l.quotaDuration)
+	l.counter = 0
+}
+
+func (l *limiter) increment() {
+	l.counter++
+}
+
+func (l *limiter) hasExpired() bool {
+	return time.Now().After(l.resetsAt)
+}
+
+func (l *limiter) hasExceededRequestQuota() bool {
+	return l.counter >= l.requestQuota
+}
+
+func (l *limiter) getRemainingRequestQuota() int {
+	return l.requestQuota - l.counter
+}
+
+func (l *limiter) getSecondsToQuotaReset() float64 {
```

Comment:
Shouldn't it return an int?

Reply:
I used float64 only because `math.Round` returns float64, and I had no straight forward reason to convert it to int.

Kindly advise

Reply:
Right, `Sprintf` with `%v` will not show the decimal places anyway if there are none.